toolchain/check/context.cpp

```diff
@@ -493,121 +493,92 @@ auto Context::ConvertToValueOrReferenceExpression(SemIR::NodeId expr_id,
   }
 }
 
-auto Context::MarkInitializerFor(SemIR::NodeId init_id, SemIR::NodeId target_id)
-    -> void {
-  SemIR::Node init = semantics_ir().GetNode(init_id);
-  CARBON_CHECK(SemIR::GetExpressionCategory(semantics_ir(), init_id) ==
-               SemIR::ExpressionCategory::Initializing)
-      << "initialization from non-initializing node " << init;
-
+// Given an initializing expression, find its return slot. Returns `Invalid` if
+// there is no return slot, because the initialization is not performed in
+// place.
+static auto FindReturnSlotForInitializer(SemIR::File& semantics_ir,
+                                         SemIR::NodeId init_id)
+    -> SemIR::NodeId {
+  SemIR::Node init = semantics_ir.GetNode(init_id);
   switch (init.kind()) {
     default:
       CARBON_FATAL() << "Initialization from unexpected node " << init;
 
     case SemIR::NodeKind::StructInit:
     case SemIR::NodeKind::TupleInit:
-    case SemIR::NodeKind::InitializeFrom:
-      CARBON_FATAL() << init << " should already have a destination";
+      // TODO: Track a return slot for these initializers.
+      CARBON_FATAL() << init
+                     << " should be created with its return slot already "
+                        "filled in properly";
+
+    case SemIR::NodeKind::InitializeFrom: {
+      auto [src_id, dest_id] = init.GetAsInitializeFrom();
+      return dest_id;
+    }
 
     case SemIR::NodeKind::Call: {
-      // If the callee has a return slot, point it at our target.
       auto [refs_id, callee_id] = init.GetAsCall();
-      if (semantics_ir().GetFunction(callee_id).return_slot_id.is_valid()) {
-        // Replace the return slot with our given target, and remove the
-        // tentatively-created temporary.
-        auto temporary_id = std::exchange(
-            semantics_ir().GetNodeBlock(refs_id).back(), target_id);
-        auto temporary = semantics_ir().GetNode(temporary_id);
-        CARBON_CHECK(temporary.kind() == SemIR::NodeKind::TemporaryStorage)
-            << "Return slot for function call does not contain a temporary; "
-            << "initialized multiple times? Have " << temporary;
-        semantics_ir().ReplaceNode(
-            temporary_id, SemIR::Node::NoOp::Make(temporary.parse_node()));
+      if (!semantics_ir.GetFunction(callee_id).return_slot_id.is_valid()) {
+        return SemIR::NodeId::Invalid;
```

Comment:
Hmm, is this going to cause the caller to restructure how the return is managed (making it not in-place)? For an invalid function, that seems like it might result in cascading errors. Should this instead return a return slot that is marked invalid? Or in some other way model that distinction?

Reply:
The situation being handled here is that the function doesn't have a return slot (which is modeled by the return slot ID being marked as invalid). Unfortunately, we fairly pervasively use `SemIr::NodeId::Invalid` to mean "none" (not a valid ID) rather than "the program is invalid". Perhaps we should rename it.